### PR TITLE
llama : sync gguf-llama with llama

### PR DESCRIPTION
**Pull request by ggerganov**

### CODE REVIEW

File: tests/test-tokenizer-1.cpp [hunk 1]
```diff
@@ -106,7 +106,12 @@ int main(int argc, char **argv) {
     std::wstring_convert<typename std::codecvt_utf8<wchar_t>, wchar_t> converter;
     for (wchar_t ch = 0x0000; ch < 0xffff; ++ch) {
         std::wstring wstr(1, ch);
-        std::string str = converter.to_bytes(wstr);
+        std::string str;
+        try {
+            str = converter.to_bytes(wstr);
+        } catch (std::exception & e) {
+            continue;
+        }
```
Comment (ggerganov) [inline tests/test-tokenizer-1.cpp:114]: @goerch For now I've added this workaround. Will try the `std::u16string` later

**Comment by ggerganov** [comment]
Any ideas why the prompt now starts with `!` instead of whitespace ` ` ?
Looks like BOS is converted to `1 + 32` -> `!`:

https://github.com/ggerganov/llama.cpp/pull/2613/files#diff-6e1ad68d87e65f7c66edf41ee2dfc5eebdbf499dfbd22fb610382373daf2b9f2R2187

Or is it because I haven't handled the special tokens yet (I'm using the new `convert-llama-h5-to-gguf.py` script) ?

**Comment by goerch** [comment]
Yep, looks like it still contains the old conversions

```
    for i in range(tokenizer.vocab_size()):
        text: bytes
        if tokenizer.is_unknown(i):
            text = " \u2047 ".encode("utf-8")
        elif tokenizer.is_control(i):
            text = b""
        if tokenizer.is_byte(i):
            piece = tokenizer.id_to_piece(i)
            if len(piece) != 6:
                raise Exception(f"Invalid token: {piece}")
            byte_value = int(piece[3:-1], 16)
            text = struct.pack("B", byte_value)
        else:
            text = tokenizer.id_to_piece(i).replace("\u2581", " ").encode("utf-8")
        score: float = tokenizer.get_score(i)

        tokens.append(text)
        scores.append(score)
```

File: tests/test-tokenizer-1.cpp (hunk 1 above)
Comment (ggerganov) [reply tests/test-tokenizer-1.cpp]: Tried it - still crashes the same way. Will keep the `try / catch` for now

**Pull request merged**